src/include/daos_obj.h

```diff
@@ -355,6 +355,54 @@ daos_obj_generate_id(daos_obj_id_t *oid, daos_ofeat_t ofeats,
 	oid->hi |= hdr;
 }
 
+typedef enum {
+	DAOS_NO_RDD,
+	DAOS_ANY_RDD,
+	DAOS_RP_RDD,
+	DAOS_EC_RDD,
+} daos_oclass_rdd_t;
+
+typedef enum {
+	DAOS_OBJ_TINY,		/** 1 grp */
+	DAOS_OBJ_SMALL,		/** ~25% */
+	DAOS_OBJ_MEDIUM,	/** ~50% */
+	DAOS_OBJ_LARGE,		/** ~75% */
+	DAOS_OBJ_MAX,		/** ~100% */
+} daos_oclass_size_t;
+
+typedef struct {
+	daos_oclass_rdd_t	oc_rdd;
+	daos_oclass_size_t	oc_size;
+} daos_oclass_hints_t;
+
+/**
+ * Generate a DAOS object ID by encoding the private DAOS bits of the object
+ * address space. This allows the user to either select an object class
+ * manually, or ask DAOS to generate one based on some hints provided.
+ *
+
+ * \param[in]	coh	Container open handle.
+ * \param[in,out]
+ *		oid	[in]: Object ID with low 96 bits set and unique inside
+ *			the container.
+ *			[out]: Fully populated DAOS object identifier with the
+ *			the low 96 bits untouched and the DAOS private bits
+ *			(the high 32 bits) encoded.
+ * \param[in]	ofeats	Feature bits specific to object
+ * \param[in]	cid	Class Identifier. If set to 0 (unkown), and hints is
+ *			NULL, OC_S1 will be selected. If one of the MAX GRP
+ *			oclass is set (SX, RP_2/3GX, etc.) the oclass chosen
+ *			will be modified to have X represent the closest GRP
+ *			available in the predefined oclass list.
+ * \param[in]   hints	Optional hints to select oclass with redundancy type
+ *			and size. This will be ignored if cid is not OC_UNKNOWN.
+ * \param[in]	args	Reserved.
+ */
+int
+daos_obj_generate_oid(daos_handle_t coh, daos_obj_id_t *oid,
+		      daos_ofeat_t ofeats, daos_oclass_id_t cid,
+		      daos_oclass_hints_t *hints, uint32_t args);
```

Comment:
Probably we should separate this function into 2 API.  
1. obj_class_id = choose_object_class(hints);
2. daos_obj_generate_oid(coh, oid, ofeats, cid, hints, args).

choose_object_class will be used later if we support other hints.  Btw: do we still need SX object class inside the stack? I would rather the SX object does not exist internally.

Reply:
2 functions would be more of a hassle IMO for users. in addition it would be extra overhead bec we still need to pass the coh in the first function to check for RF property too.
we still need SX in the stack to maintain bw compatibility with others. we also need to add something like OC_SXX for benchmarks to be able to quickly select the max sharding.